File: src/test/java/net/pterodactylus/sone/TestPostBuilder.java
package net.pterodactylus.sone;

import static com.google.common.base.Optional.fromNullable;
import static java.lang.System.currentTimeMillis;
import static java.util.UUID.randomUUID;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import net.pterodactylus.sone.data.Post;
import net.pterodactylus.sone.data.Sone;
import net.pterodactylus.sone.database.PostBuilder;

/**
 * {@link PostBuilder} implementation that returns a mocked {@link Post}.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class TestPostBuilder implements PostBuilder {

	private final Post post = mock(Post.class);
	private String recipientId = null;

	@Override
	public PostBuilder copyPost(Post post) throws NullPointerException {
		return this;
	}

	@Override
	public PostBuilder from(String senderId) {
		final Sone sone = mock(Sone.class);
		when(sone.getId()).thenReturn(senderId);
		when(post.getSone()).thenReturn(sone);
		return this;
	}

	@Override
	public PostBuilder randomId() {
		when(post.getId()).thenReturn(randomUUID().toString());
		return this;
	}

	@Override
	public PostBuilder withId(String id) {
		when(post.getId()).thenReturn(id);
		return this;
	}

	@Override
	public PostBuilder currentTime() {
		when(post.getTime()).thenReturn(currentTimeMillis());
		return this;
	}

	@Override
	public PostBuilder withTime(long time) {
		when(post.getTime()).thenReturn(time);
		return this;
	}

	@Override
	public PostBuilder withText(String text) {
		when(post.getText()).thenReturn(text);
		return this;
	}

	@Override
	public PostBuilder to(String recipientId) {
		this.recipientId = recipientId;
		return this;
	}

	@Override
	public Post build() throws IllegalStateException {
		when(post.getRecipientId()).thenReturn(fromNullable(recipientId));
		return post;
	}

}

File: src/test/kotlin/net/pterodactylus/sone/test/TestPostBuilder.kt
package net.pterodactylus.sone.test

import com.google.common.base.Optional
import net.pterodactylus.sone.data.Post
import net.pterodactylus.sone.data.Sone
import net.pterodactylus.sone.database.PostBuilder
import java.util.UUID

/**
 * [PostBuilder] implementation that returns a mocked [Post].
 */
class TestPostBuilder : PostBuilder {

	private val post = mock<Post>()
	private var recipientId: String? = null

	override fun copyPost(post: Post): PostBuilder = this

	override fun from(senderId: String): PostBuilder = apply {
		val sone = mock<Sone>()
		whenever(sone.id).thenReturn(senderId)
		whenever(post.sone).thenReturn(sone)
	}

	override fun randomId(): PostBuilder = apply {
		whenever(post.id).thenReturn(UUID.randomUUID().toString())
	}

	override fun withId(id: String): PostBuilder = apply {
		whenever(post.id).thenReturn(id)
	}

	override fun currentTime(): PostBuilder = apply {
		whenever(post.time).thenReturn(System.currentTimeMillis())
	}

	override fun withTime(time: Long): PostBuilder = apply {
		whenever(post.time).thenReturn(time)
	}

	override fun withText(text: String): PostBuilder = apply {
		whenever(post.text).thenReturn(text)
	}

	override fun to(recipientId: String): PostBuilder = apply {
		this.recipientId = recipientId
	}

	override fun build(): Post = post
			.also {
				whenever(post.recipientId).thenReturn(Optional.fromNullable(recipientId))
			}

}
